x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/ResultsIndexUpgradeAction.java

```diff
@@ -0,0 +1,158 @@
+
+import org.elasticsearch.action.Action;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.IndicesRequest;
+import org.elasticsearch.action.support.IndicesOptions;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.action.support.master.MasterNodeReadOperationRequestBuilder;
+import org.elasticsearch.action.support.master.MasterNodeReadRequest;
+import org.elasticsearch.client.ElasticsearchClient;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.tasks.CancellableTask;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.xpack.core.ml.job.persistence.AnomalyDetectorsIndexFields;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+
+public class ResultsIndexUpgradeAction extends Action<AcknowledgedResponse> {
+    public static final ResultsIndexUpgradeAction INSTANCE = new ResultsIndexUpgradeAction();
+    public static final String NAME = "cluster:admin/xpack/ml/results/upgrade";
+
+    private ResultsIndexUpgradeAction() {
+        super(NAME);
+    }
+
+    @Override
+    public AcknowledgedResponse newResponse() {
+        return new AcknowledgedResponse();
+    }
+
+    public static class Request
+        extends MasterNodeReadRequest<ResultsIndexUpgradeAction.Request>
+        implements IndicesRequest, ToXContentObject {
+
+        private static final ParseField REINDEX_BATCH_SIZE = new ParseField("reindex_batch_size");
+
+        public static ObjectParser<Request, Void> PARSER = new ObjectParser<>("ml_results_index_upgrade", true, Request::new);
+        static {
+            PARSER.declareInt(Request::setReindexBatchSize, REINDEX_BATCH_SIZE);
+        }
+
+        static final String INDEX = AnomalyDetectorsIndexFields.RESULTS_INDEX_PREFIX + "*";
+        private int reindexBatchSize = 1000;
+
+        /**
+         * Should this task store its result?
+         */
+        private boolean shouldStoreResult;
+
+        // for serialization
+        public Request() {
+        }
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            reindexBatchSize = in.readInt();
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeInt(reindexBatchSize);
+        }
+
+        @Override
+        public String[] indices() {
+            return new String[]{INDEX};
+        }
+
+        @Override
+        public IndicesOptions indicesOptions() {
+            //TODO consider lenientExpandOpen() ?
```

Comment:
I am not sure of the best index option when opening up the `.ml-anomalies-*` index request. I thought that disallowing `closed` made the most sense.

Reply:
I agree it's probably best to fail the entire request if there are closed indices that need upgrading.  This is extremely unlikely to happen, as we never close ML internal indices and don't advise users to.  But it would be good to confirm that the error message is clear if this endpoint is called when an ML results index is closed.